src/librustc_mir/interpret/memory.rs

```diff
@@ -414,8 +414,15 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
         id: AllocId,
         is_write: bool,
     ) -> InterpResult<'tcx, Cow<'tcx, Allocation<M::PointerTag, M::AllocExtra>>> {
-        let alloc = tcx.alloc_map.lock().get(id);
-        let (alloc, def_id) = match alloc {
+        // The call to `resolve_maybe_global_alloc` is needed to enable Miri to
+        // support thread local statics. In `M::canonical_alloc_id`, for a
+        // thread local static, Miri reserves a fresh allocation id, but the
+        // actual allocation is left to the code that handles statics which
+        // calls this function (`get_global_alloc`). Since the allocation id is
+        // fresh, it has no information about the original static. The call to
+        // `resolve_maybe_global_alloc` allows Miri to retrieve this information
+        // for us.
+        let (alloc, def_id) = match M::resolve_maybe_global_alloc(tcx, memory_extra, id) {
```

Comment:
This seems quite roundabout... so given the lazy `AllocId` of a thread_local static (the one that resolves to a `DefId`), we first call `canonical_alloc_id` to map this *away*, to a "real" `AllocId`, and then here we have to map it *back*.

I think we should avoid that round-trip. I will think about how to do that, but input is welcome.

Reply:
I think whatever we do here is heavily affected by https://github.com/rust-lang/rust/pull/70598/files#r401254938, so I won't spend more time here until I understand how that works.^^

Reply:
> I think we should avoid that round-trip. I will think about how to do that, but input is welcome.

I listed the three alternatives I considered in [this comment](https://github.com/rust-lang/rust/pull/70598/files#r401327477).

The current approach while definitely not ideal, feels the least intrusive and, therefore, the most acceptable to me until we have  #70685 fixed.